src/core/manager.c

```diff
@@ -2986,10 +2995,49 @@ void manager_check_finished(Manager *m) {
         manager_invalidate_startup_units(m);
 }
 
+static bool generator_path_any(const char* const* paths) {
+        char **path;
+        bool found = false;
+
+        /* Optimize by skipping the whole process by not creating output directories
+         * if no generators are found. */
+        STRV_FOREACH(path, (char**) paths)
+                if (access(*path, F_OK) == 0)
+                        found = true;
+                else if (errno != ENOENT)
+                        log_warning_errno(errno, "Failed to open generator directory %s: %m", *path);
+
+        return found;
+}
+
+static const char* environment_generator_binary_paths[] = {
+        "/run/systemd/environment-generators",
+        "/etc/systemd/environment-generators",
+        "/usr/local/lib/systemd/environment-generators",
+        ENVIRONMENT_GENERATOR_PATH,
+        NULL
+};
+
+static int manager_run_environment_generators(Manager *m) {
+        char **tmp = NULL; /* this is only used in the forked process, no cleanup here */
+        const gather_stdout_callbacks cb = {
+                gather_environment_one, &tmp,
+                gather_environment_two, &tmp,
+                gather_environment_three, &m->environment,
+        };
+
+        if (!generator_path_any(environment_generator_binary_paths))
+                return 0;
```

Comment:
hmm, so what's the story here? this is called for both `--system` and `--user` mode the way this is right now? Is this really the intention?

If this is the intention, shouldn't the dirs be called "system-environment-generators" and "user-environment-generators"? If that's not the intention, shouldn't the dir be called "user-environment-generators" anyway, simply that the door is open to maybe one day add this for the system generators too? (not saying htat would be a good idea, just want to keep the avenue open...) or am i missing something?